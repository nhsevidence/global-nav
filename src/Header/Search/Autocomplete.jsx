import React, { Component } from "react";
import Autocomplete from "accessible-autocomplete/react";

import "!style-loader!css-loader!accessible-autocomplete/dist/accessible-autocomplete.min.css";

function suggest(query, syncResults) {
	fetch(`https://www.nice.org.uk/autocomplete?q=${query}&ajax=ajax`)
		.then(response => response.json())
		.then(data => {
			syncResults(data);
		});
}

const templates = {
	inputValue: suggestion => suggestion && suggestion.Title,
	suggestion: suggestion =>
		suggestion && `<a href="${suggestion.Link}">${suggestion.Title}</a>`
};

const onConfirm = suggestion => {
	if (suggestion) window.location.href = suggestion.Link;
};

export default class Search extends Component {
	render() {
		return (
			<Autocomplete
				id="autocomplete"
				source={suggest}
				templates={templates}
				onConfirm={onConfirm}
			/>
		);
	}
}
