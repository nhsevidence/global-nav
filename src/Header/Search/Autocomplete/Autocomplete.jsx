import React, { Component } from "react";
import AccessibleAutocomplete from "accessible-autocomplete/react";
import PropTypes from "prop-types";
import styles from "./Autocomplete.module.scss";

import { suggester } from "./suggester";

import { trackEvent } from "./../../../tracker";

/**
 * Debounce
 * See http://unscriptable.com/2009/03/20/debouncing-javascript-methods/
 *
 * @param      {Function}  func       The function to debounce
 * @param      {Integer}  execAsap   Whether to execute the function now
 * @param      {Integer}  threshold  The detection period, in milliseconds
 * @param      {Object}  scope  The context for the debounced function
 * @return     {Function}  { The debounced function }
 */
const debounce = function(
	func,
	execAsap = false,
	threshold = 100,
	scope = null
) {
	let timeout;

	return function debounced() {
		let context = scope || this,
			args = arguments;

		function delayed() {
			if (!execAsap) func.apply(context, args);
			timeout = null;
		}

		if (timeout) clearTimeout(timeout);
		else if (execAsap) func.apply(context, args);

		timeout = setTimeout(delayed, threshold);
	};
};

/** The maximum number of autocomplete results to return */
const maxResults = 5;

/** Delay in millieconds before loading results */
export const rateLimitWait = 300;

const templates = {
	inputValue: function(suggestion) {
		return suggestion && suggestion.Title;
	},
	suggestion: function(suggestion) {
		return (
			suggestion &&
			`<a href="${suggestion.Link}">${suggestion.TitleHtml ||
				suggestion.Title}</a>`
		);
	}
};

const onConfirm = function(suggestion) {
	if (suggestion) {
		const eventCallback = function() {
			window.location.href = suggestion.Link;
		};
		if (suggestion.TypeAheadType != undefined) {
			trackEvent(
				"Search - Typeahead select",
				"Selected: " + suggestion.TypeAheadType,
				(
					suggestion.Title +
					" | " +
					document.getElementById("autocomplete").value
				).toLowerCase(),
				null,
				eventCallback
			);
		} else {
			trackEvent(
				"Search",
				"Typeahead select",
				suggestion.Title +
					" | " +
					document.getElementById("autocomplete").value,
				null,
				eventCallback
			);
		}
	}
};

export default class Autocomplete extends Component {
	suggest(query, populateResults) {
		if (
			this.props.source === false ||
			query === "" ||
			query === this.props.placeholder
		)
			return;

		let source;

		if (Array.isArray(this.props.source)) {
			source = this.props.source;
		}

		if (
			typeof this.props.source === "string" &&
			this.props.source.indexOf("/") === -1
		) {
			source = window[this.props.source];
			if (!source) return;
		}

		if (source) {
			populateResults(suggester(source, query, maxResults));
			return;
		}

		// Default to a URL for asynchronously loading suggestions from the server
		fetch(
			`${this.props.source}${
				this.props.source.indexOf("?") === -1 ? "?" : "&"
			}q=${query}`
		)
			.then(response => response.json())
			.then(data => {
				populateResults(data.slice(0, maxResults));
			});
	}

	render() {
		const isIE8 = function() {
			if (typeof navigator === "undefined") return false; // For server rendering
			const ua = navigator.userAgent;
			var msie = ua.indexOf("MSIE ");
			if (msie > 0) {
				return parseInt(ua.substring(msie + 5, ua.indexOf(".", msie)), 10) <= 8;
			}

			return false;
		};

		return (
			<div className={styles.ac}>
				{!this.props.source || isIE8() || typeof window === "undefined" ? (
					<div className="autocomplete__wrapper">
						<input
							type="search"
							id="autocomplete"
							name="q"
							className="autocomplete__input autocomplete__input--default"
							placeholder={this.props.placeholder}
							defaultValue={this.props.query}
							data-hj-whitelist=""
						/>
					</div>
				) : (
					<AccessibleAutocomplete
						id="autocomplete"
						name="q"
						placeholder={this.props.placeholder}
						displayMenu="overlay"
						minLength={3}
						source={debounce(this.suggest, false, rateLimitWait, this)}
						templates={templates}
						onConfirm={onConfirm}
						confirmOnBlur={false}
						showNoOptionsFound={false}
						defaultValue={this.props.query}
						ref={function(acElement) {
							// TODO: This relies on an inner implementation detail of the autocomplete component, can we do this in a better way?
							acElement &&
								acElement.elementReferences &&
								acElement.elementReferences[-1] &&
								acElement.elementReferences[-1].setAttribute(
									"data-hj-whitelist",
									""
								);
						}}
					/>
				)}
			</div>
		);
	}
}

Autocomplete.propTypes = {
	source: PropTypes.oneOfType([
		PropTypes.bool,
		PropTypes.string,
		PropTypes.arrayOf(
			PropTypes.shape({ Title: PropTypes.string, Link: PropTypes.string })
		)
	]),
	placeholder: PropTypes.string,
	query: PropTypes.string
};

Autocomplete.defaultProps = {
	source: false
};
