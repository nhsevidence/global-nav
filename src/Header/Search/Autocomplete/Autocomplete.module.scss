@import '~@nice-digital/nds-core/scss/core';

/* stylelint-disable */

@import '../../header-settings';

.ac {
  :global {
    .autocomplete__wrapper {
      position: relative;
    }

    .autocomplete__hint,
    .autocomplete__input {
      -webkit-appearance: none;
      border: 1px solid $colour-nice-grey-2;
      border-radius: 0; /* Safari 10 on iOS adds implicit border rounding. */
      box-sizing: border-box;
      margin-bottom: 0; /* BUG: Safari 10 on macOS seems to add an implicit margin. */
      width: 100%;
    }

    .autocomplete__input {
      // The importants here are to override conflicting global styles in niceorg
      background-color: $colour-nice-grey-3 !important;
      height: rem($search-height) !important;
      max-width: 100% !important;
      position: relative;
      z-index: 101;

      &:focus {
        outline: 3px solid $colour-focus;
        outline-offset: -3px;
        z-index: 1;
      }

      &::placeholder {
        font-size: 100%;
        font-weight: normal;
      }
    }

    .autocomplete__hint {
      color: #bfc1c3;
      position: absolute;
    }

    .autocomplete__input--default {
      padding: rem($spacing-small);
    }

    .autocomplete__input--focused {
      outline: 3px solid $colour-focus;
      outline-offset: -3px;
    }

    .autocomplete__input--show-all-values {
      padding: 4px 34px 4px 4px;
      cursor: pointer;
    }

    .autocomplete__dropdown-arrow-down {
      z-index: -1;
      display: inline-block;
      position: absolute;
      right: 8px;
      width: 24px;
      height: 24px;
      top: 10px;
    }

    .autocomplete__menu {
      background-color: #fff;
      border: 1px solid $colour-nice-grey-2;
      border-top: 0;
      color: #34384b;
      margin: 0;
      max-height: 342px;
      //overflow-x: hidden;
      padding: 0;
      text-align: left;
      width: 100%;
    }

    .autocomplete__menu--visible {
      display: block;
    }

    .autocomplete__menu--hidden {
      display: none;
    }

    .autocomplete__menu--overlay {
      box-shadow: rem(0 $spacing-x-small $spacing-medium $spacing-x-small)
        rgba(0, 0, 0, 0.1);
      left: 0;
      position: absolute;
      top: 100%;
      z-index: 100;
    }

    .autocomplete__menu--inline {
      position: relative;
    }

    .autocomplete__option {
      border-bottom: solid $colour-nice-grey-2;
      border-width: 1px 0;
      color: $colour-link;
      cursor: pointer;
      display: block;
      margin: 0;
      position: relative;

      a {
        color: inherit;
        display: block;
        font-weight: normal;
        padding: rem($spacing-small);
        text-decoration: none;
      }

      mark {
        background: 0;
        color: inherit;
        font-weight: bold;
        padding: 0;
      }
    }

    .autocomplete__option > * {
      pointer-events: none;
    }

    .autocomplete__option:first-of-type {
      border-top-width: 0;
    }

    .autocomplete__option:last-of-type {
      border-bottom-width: 0;
    }

    .autocomplete__option--odd {
      background-color: #fafafa;
    }

    .autocomplete__option--focused,
    .autocomplete__option:hover {
      background-color: $colour-nice-grey-3;
      color: $colour-link;
      outline: none;
      text-decoration: underline;
    }

    .autocomplete__option--no-results {
      background-color: #fafafa;
      color: #646b6f;
      cursor: not-allowed;
    }

    .autocomplete__hint,
    .autocomplete__input,
    .autocomplete__option {
      font-size: 16px;
      line-height: 1.25;
    }

    .autocomplete__hint,
    .autocomplete__option {
      //padding: 4px;
    }
  }
}
