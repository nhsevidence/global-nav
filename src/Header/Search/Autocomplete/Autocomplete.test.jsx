import React from "react";
import AccessibleAutocomplete from "accessible-autocomplete/react";

import Autocomplete from "./Autocomplete";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";

import { eventName, eventTimeout } from "./../../../tracker";

const IE8UserAgent =
	"Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 6.0; Trident/4.0)";

describe("Autocomplete", () => {
	const defaultProps = {};

	describe("Rendering", () => {
		it("Renders without crashing", () => {
			const wrapper = shallow(<Autocomplete {...defaultProps} />);
			expect(wrapper).toHaveLength(1);
		});
	});

	describe("Autocomplete", () => {
		it("should render autocomplete component", () => {
			const wrapper = shallow(
				<Autocomplete {...defaultProps} source="/url" query="test" />
			);
			expect(toJson(wrapper)).toMatchSnapshot();
		});

		it("should render query into defaultValue attribute on autocomplete component", () => {
			const wrapper = shallow(
				<Autocomplete {...defaultProps} source="/url" query="diabetes" />
			);
			expect(wrapper.find(AccessibleAutocomplete).props().defaultValue).toEqual(
				"diabetes"
			);
		});

		it("should add HotJar whitelist attribute to input box", () => {
			const wrapper = mount(
				<Autocomplete {...defaultProps} source="/url" query="diabetes" />
			);

			expect(
				wrapper
					.getDOMNode()
					.querySelector("input#autocomplete")
					.getAttribute("data-hj-whitelist")
			).toEqual("");
		});

		it("should push autocomplete select event to the dataLayer", () => {
			document.body.innerHTML = "";
			var appContainer = document.createElement("div");
			document.body.appendChild(appContainer);

			const option = {
				Title: "diabetes type 1",
				Link: "https://www.nice.org.uk/diabetes1.html"
			};

			const wrapper = mount(
				<Autocomplete {...defaultProps} source={[option]} query="diab" />,
				{ attachTo: appContainer }
			);

			wrapper
				.find("#autocomplete")
				.first()
				.props()
				.onConfirm(option);

			expect(window.dataLayer).toEqual([
				{
					event: eventName,
					eventCategory: "Search",
					eventAction: "Typeahead select",
					eventLabel: option.Title + " | diab",
					eventCallback: expect.any(Function),
					eventTimeout: eventTimeout
				}
			]);
		});

		it("should navigate to selected option in event callback", () => {
			document.body.innerHTML = "";
			var appContainer = document.createElement("div");
			document.body.appendChild(appContainer);

			const option = {
				Title: "diabetes type 1",
				Link: "https://www.nice.org.uk/diabetes1.html"
			};

			const wrapper = mount(
				<Autocomplete {...defaultProps} source={[option]} query="diab" />,
				{ attachTo: appContainer }
			);

			wrapper
				.find("#autocomplete")
				.first()
				.props()
				.onConfirm(option);

			window.dataLayer[0].eventCallback();
			expect(window.location.href).toEqual(
				"https://www.nice.org.uk/diabetes1.html"
			);
		});
	});

	describe("Non-autocomplete", () => {
		it("should render fallback search box when no source", () => {
			const wrapper = shallow(
				<Autocomplete {...defaultProps} source={false} query="diabetes" />
			);
			expect(toJson(wrapper)).toMatchSnapshot();
		});

		it("should render fallback search box in IE8", () => {
			const oldNavigator = global.navigator;
			navigator.userAgent = IE8UserAgent;

			const wrapper = shallow(
				<Autocomplete
					{...defaultProps}
					source="/a-valid-source"
					query="diabetes"
				/>
			);
			expect(toJson(wrapper)).toMatchSnapshot();

			navigator.userAgent = oldNavigator;
		});
	});
});
