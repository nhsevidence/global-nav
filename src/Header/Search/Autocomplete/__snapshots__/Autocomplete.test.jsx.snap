// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete Autocomplete should render autocomplete component 1`] = `
<div
  className="ac"
>
  <e
    autoselect={false}
    confirmOnBlur={false}
    cssNamespace="autocomplete"
    defaultValue="test"
    displayMenu="overlay"
    dropdownArrow={[Function]}
    id="autocomplete"
    minLength={3}
    name="q"
    onConfirm={[Function]}
    placeholder=""
    required={false}
    showAllValues={false}
    showNoOptionsFound={false}
    source={[Function]}
    tAssistiveHint={[Function]}
    tNoResults={[Function]}
    templates={
      Object {
        "inputValue": [Function],
        "suggestion": [Function],
      }
    }
  />
</div>
`;

exports[`Autocomplete Non-autocomplete should render fallback search box when no source 1`] = `
<div
  className="ac"
>
  <div
    className="autocomplete__wrapper"
  >
    <input
      className="autocomplete__input autocomplete__input--default"
      data-hj-allow=""
      defaultValue="diabetes"
      id="autocomplete"
      maxLength={512}
      name="q"
      type="search"
    />
  </div>
</div>
`;
